store/tikv/client.go

```diff
@@ -15,140 +15,232 @@
 	return &rpcClient{
-		msgID: 0,
-		p: NewPools(maxConnection, func(addr string) (*Conn, error) {
-			return NewConnection(addr, dialTimeout)
-		}),
+		conns: make(map[string]*grpc.ClientConn),
 	}
 }
 
-// SendCopReq sends a Request to co-processor and receives Response.
-func (c *rpcClient) SendCopReq(ctx goctx.Context, addr string, req *coprocessor.Request, timeout time.Duration) (*coprocessor.Response, error) {
-	start := time.Now()
-	defer func() { sendReqHistogram.WithLabelValues("cop").Observe(time.Since(start).Seconds()) }()
-
-	select {
-	case <-ctx.Done():
-		return nil, ctx.Err()
-	default:
-	}
-
-	conn, err := c.p.GetConn(addr)
-	if err != nil {
-		return nil, errors.Trace(err)
+func (c *rpcClient) getConn(addr string) (*grpc.ClientConn, error) {
+	c.RLock()
+	if c.isClosed {
+		c.RUnlock()
+		return nil, errors.Errorf("rpcClient is closed")
 	}
-	connPoolHistogram.WithLabelValues("cop").Observe(time.Since(start).Seconds())
-	defer c.p.PutConn(conn)
-	msg := msgpb.Message{
-		MsgType: msgpb.MessageType_CopReq,
-		CopReq:  req,
+	conn, ok := c.conns[addr]
+	c.RUnlock()
+	if !ok {
+		var err error
+		conn, err = c.createConn(addr)
+		if err != nil {
+			return nil, err
+		}
 	}
-	err = c.doSend(conn, &msg, writeTimeout, timeout)
-	if err != nil {
-		conn.Close()
-		return nil, errors.Trace(err)
+	return conn, nil
+}
+
+func (c *rpcClient) createConn(addr string) (*grpc.ClientConn, error) {
+	c.Lock()
+	defer c.Unlock()
+	conn, ok := c.conns[addr]
+	if !ok {
+		var err error
+		conn, err = grpc.Dial(
+			addr,
+			grpc.WithInsecure(),
+			grpc.WithTimeout(dialTimeout),
+			grpc.WithUnaryInterceptor(grpc_prometheus.UnaryClientInterceptor),
+			grpc.WithStreamInterceptor(grpc_prometheus.StreamClientInterceptor))
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		c.conns[addr] = conn
 	}
-	if msg.GetMsgType() != msgpb.MessageType_CopResp || msg.GetCopResp() == nil {
-		conn.Close()
-		return nil, errors.Trace(errInvalidResponse)
+	return conn, nil
+}
+
+func (c *rpcClient) closeConns() {
+	c.Lock()
+	if !c.isClosed {
+		c.isClosed = true
+		// close all connections
+		for _, conn := range c.conns {
+			conn.Close()
+		}
 	}
-	return msg.GetCopResp(), nil
+	c.Unlock()
 }
 
-// SendKVReq sends a Request to kv server and receives Response.
-func (c *rpcClient) SendKVReq(ctx goctx.Context, addr string, req *kvrpcpb.Request, timeout time.Duration) (*kvrpcpb.Response, error) {
+// SendReq sends a Request to server and receives Response.
+func (c *rpcClient) SendReq(ctx goctx.Context, addr string, req *tikvrpc.Request) (*tikvrpc.Response, error) {
 	start := time.Now()
-	defer func() { sendReqHistogram.WithLabelValues("kv").Observe(time.Since(start).Seconds()) }()
-
-	select {
-	case <-ctx.Done():
-		return nil, ctx.Err()
-	default:
+	var label = rpcLabelKV
+	if req.Type == tikvrpc.CmdCop {
+		label = rpcLabelCop
 	}
+	defer func() { sendReqHistogram.WithLabelValues(label).Observe(time.Since(start).Seconds()) }()
```

Comment:
The grpc metrics calculate the sum time between sending requests and receiving response.
The metric here is to record the time it pushes request to grpc layer.